sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/MetricExporterFactory.java

```diff
@@ -49,93 +43,39 @@ public MetricExporter create(
 
     OtlpMetric otlpModel = model.getOtlp();
     if (otlpModel != null) {
-      return FileConfigUtil.addAndReturn(closeables, createOtlpExporter(otlpModel, spiHelper));
+      model.getAdditionalProperties().put("otlp", otlpModel);
     }
 
     if (model.getConsole() != null) {
-      return FileConfigUtil.addAndReturn(closeables, createConsoleExporter(spiHelper));
+      model.getAdditionalProperties().put("console", model.getConsole());
     }
 
     if (model.getPrometheus() != null) {
       throw new ConfigurationException("prometheus exporter not supported in this context");
     }
 
-    // TODO(jack-berg): add support for generic SPI exporters
     if (!model.getAdditionalProperties().isEmpty()) {
-      throw new ConfigurationException(
-          "Unrecognized metric exporter(s): "
-              + model.getAdditionalProperties().keySet().stream().collect(joining(",", "[", "]")));
+      Map<String, Object> additionalProperties = model.getAdditionalProperties();
+      if (additionalProperties.size() > 1) {
+        throw new ConfigurationException(
+            "Invalid configuration - multiple metric exporters set: "
+                + additionalProperties.keySet().stream().collect(joining(",", "[", "]")));
+      }
+      Map.Entry<String, Object> exporterKeyValue =
+          additionalProperties.entrySet().stream()
+              .findFirst()
+              .orElseThrow(
+                  () ->
+                      new IllegalStateException("Missing exporter. This is a programming error."));
+      MetricExporter metricExporter =
+          FileConfigUtil.loadComponent(
+              spiHelper,
+              MetricExporter.class,
+              exporterKeyValue.getKey(),
+              exporterKeyValue.getValue());
+      return FileConfigUtil.addAndReturn(closeables, metricExporter);
     }
 
     return null;
```

Comment:
Is returning nulls from these Factory implementations a pattern that we like? I looked at the interface definition, and it didn't seem to imply that null return value was acceptable.

```
interface Factory<ModelT, ResultT> {

  /**
   * Interpret the model and create {@link ResultT} with corresponding configuration.
   *
   * @param model the configuration model
   * @param spiHelper the service loader helper
   * @param closeables mutable list of closeables created
   * @return the {@link ResultT}
   */
  ResultT create(@Nullable ModelT model, SpiHelper spiHelper, List<Closeable> closeables);
}
```

Reply:
I think the contract of the parent `Factory` interface should probably be adjusted. 

Currently, the contract is `ResultT create(@Nullable ModelT model, SpiHelper spiHelper, List<Closeable> closeables)`, so it can accept a null model but is always expected to either return (or throw). This is odd. Always returning when the model is null means that every type has to have some no-op or default representation, which is too strict. Throwing when the model is nullable is a weird pattern too. Why allow the model to be nullable if its just going to throw NPE?

So I think we (i.e. me) need to adjust the factory contract to be consistent: Either both the input model and response are nullabe, OR neither is nullable. And it seems best if neither the input or the response is null. This means that callers have to do some additional work to ensure they're not working with a null model, but should be a good thing in net. 

If you're on board, I'll open up an issue and follow up. For the purposes of this PR, there is no appropriate MetricExporter to return when the input model is null, so we need to throw an exception or return null. Returning null is preferable because it is handled by the caller and we can generate a more appropriate error message in `MetricReaderFactory` than we can in `MetricExporterFactory`:

https://github.com/open-telemetry/opentelemetry-java/blob/9a13f6880d53a19344d0e49dbd6c7daadbbab68a/sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/MetricReaderFactory.java#L49-L53

Reply:
Seems totally reasonable to me to adjust the interface to reflect what appears to be the reality.  The other Option (heh) would be to return `Optional<ResultT>` from the interface, which might signal the correct semantics? Since this isn't a hot-path interface, I think we should at least consider it.

Reply:
Created #6612 to refactor away from using nullable in the factory contract. Not sure why I thought that was necessary, but it was clearly a mistake.

Reply:
Any thoughts on using Optional?

Reply:
The need for optional goes away when we embrace avoiding `null`. Once you eliminate the ability for `Factory` to accept a `@Nullable` model, there aren't actually any cases where its useful to return `Optional.empty()`. When model is always non-null, `Factory` should always returns a valid component or throws `ConfigurationException`.

Reply:
Doh! I completely misread what you did over in the other PR. Thanks! Great refactoring of that!